### ISSUE TITLE
Local resampling trick

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#4102]
This [recent preprint](https://arxiv.org/abs/2305.05475) (from @proteneer @maxentile, and others) shows a really nice method that can greatly accelerate sampling of regions of interest. The basic idea is that, if there is an area of particular interest (e.g. ligand bound in a protein active site), you can alternate between normal "global" MD and much less expensive focused "local" MD that focuses on the region of interest. This is a two step process which involves stochastically selecting which M atoms to fix positions (where M << N atoms) and then impose a restraint based on this function during local dynamics:
![image](https://github.com/openmm/openmm/assets/3656088/dea61c07-f175-4436-97f0-bcb86e445d70)
This can speed up sampling in the region of interest by many times if the implementation supports efficient computation of the [M(M-1)/2 + MN] interactions. This is essentially what interaction groups is used for in OpenMM `CustomNonbonded` forces, but I'm not certain this feature supports pairlists yet.
![image](https://github.com/openmm/openmm/assets/3656088/068593fe-e23b-499c-9e42-daf60015ef6f)
The manuscript goes to extensive lengths to validate this approach and show that it gives equivalent results at much less cost.

**Comment by maxentile** [comment]
Thanks for the highlight, @jchodera!

Tagging @badisa , who developed the implementation in timemachine and performed much of the investigation reported in the preprint.

--------------

A "primitive" that might accelerate approaches in this vein -- including prior methods such as [[Giovannelli et al 2016]](https://pubs.acs.org/doi/pdf/10.1021/acs.jctc.5b01117) -- is, roughly, a function `force(x, selection)` that efficiently computes `force(x)[selection]`.

We might randomize the `selection` many times during the simulation, and, for a fixed `selection`, we might call the function many times on inputs where only some of the particle positions `x[selection]` vary.

(Efficient implementation of this sort of function seems clearer for systems where the nonbonded interactions have a cutoff. Extension to PME systems may be possible, following @craabreu 's nonbonded slicing plug-in https://github.com/craabreu/openmm-nonbonded-slicing . Also see discussion in a related feature request: https://github.com/openmm/openmm/issues/3971 .)

--------------

We did not yet attempt to implement this method as an OpenMM CustomIntegrator, and we have not benchmarked the use of OpenMM InteractionGroups in this setting, but it could be a natural fit. (Also, we drew on OpenMM's InteractionGroup in the design of an InteractionGroup feature in timemachine.)

In the preprint, we do mention a specific sequence of implementation choices, although I don't think I would advocate for all of them in all circumstances. (E.g. the choice to use a flat-bottom quartic restraint -- it may be appropriate to consider other flat-bottom restraints, barrier functions, non-spherical restraints, etc. -- we just haven't explored many options here yet.)

For reference, a minimal (but slow!) Jax implementation that does not commit to many of these choices is available here: https://github.com/proteneer/timemachine/pull/785

**Comment by peastman** [comment]
Can you elaborate on what `force(x, selection)` would compute?  How is it defined?  I'm assuming that `x` is the vector of particle coordinates.  Are you computing the forces on `selection` due to all particles, or the forces on all particles due to `selection`?  Does this only apply to nonbonded forces, or bonded ones as well?  If the latter, what happens when a bond involves both selected and unselected particles?

**Comment by maxentile** [comment]
>Can you elaborate on what force(x, selection) would compute? How is it defined?
>I'm assuming that x is the vector of particle coordinates. 

Yes, sorry for unclarity!

`force(x, selection) = force(x)[selection]`, where:
* `x` is the vector of particle coordinates,
* `F = force(x)` computes all forces on all particles (`F.shape == x.shape`), and 
* `selection` is a particle selection mask (boolean vector of length `len(x)`).

Without special performance optimizations, one could compute the forces on all particles `F = force(x)` and access a selection of these `F[selection]`.

But in some cases, performance optimizations may be possible to compute the same result `F[selection]` with substantially less work than computing all of `F`.

(In one iteration of the proposed method, we would make a random choice of `selection`, then compute `force(x, selection)` (+ a much cheaper restraint force) many times on inputs where (1) `selection` is fixed and (2) only some particle positions `x[selection]` are varying.)

>Are you computing the forces on selection due to all particles, or the forces on all particles due to selection?

All particles.

>Does this only apply to nonbonded forces, or bonded ones as well?

All forces.

>If the latter, what happens when a bond involves both selected and unselected particles?

That bond is included.

**Comment by jchodera** [comment]
A `CustomIntegrator` operation `force(particle_selection)` that would compute only the forces (or energies) on the atomic subset would be very useful for both multiple timestep integrators (e.g. solute/solvent splitting) and the local resampling trick. As @maxentile notes, both PME and constraints bridging the selection could be a challenge, but pairwise interactions should be straightforward. Making it efficient---especially if the `particle_selection` changes frequently---could be difficult, but there may be a clever way to do this. It would presumably have to raise an error if used with PME or other features that would cause issues.

It may be easier to consider adding this capability to `CustomNonbondedForce` since it _only_ deals with pairwise interactions. One could imagine extending the API to restrict computation of energies and forces only to a subset of the atoms (interacting with everything else), e.g.:
```python
force.setActiveParticles(atom_indices)
```
This would allow for `CustomNonbondedForce` to make use of efficient automatic pairlists for only the active particles, allowing many steps to be taken for local resampling before the set of active particles is changed and pairlist rebuilds are forced.

**Comment by peastman** [comment]
What you're talking about is just an optimization, right?  It's exactly the same as calculating the forces normally, then ignoring the particles you don't care about.  The question is whether there's a potential speedup if you don't care about all particles.  Is that correct?

Direct space nonbonded interactions probably have a potential for some speedup.  Reciprocal space interactions have very little room for speedup, since charge spreading and FFTs would have to be done normally regardless of how many particles you wanted forces on.  In principle bonded interactions could benefit, but realizing that benefit might be challenging.  You would either have to do extra checks for each interaction, which would probably take as much time as just computing it, or else you would have to build a filtered list of interactions every time the selection changed.

**Comment by jchodera** [comment]
> What you're talking about is just an optimization, right? It's exactly the same as calculating the forces normally, then ignoring the particles you don't care about. The question is whether there's a potential speedup if you don't care about all particles. Is that correct?

Yes. There is clearly a substantial speedup possible for pairwise potentials (such as reaction field), as noted in the linked preprint. The question is what would need to be done in order to ensure OpenMM could achieve this speedup for pairwise potentials---specifically for the case where multiple sequential evaluations with the same particle subset are made.

> Direct space nonbonded interactions probably have a potential for some speedup. Reciprocal space interactions have very little room for speedup, since charge spreading and FFTs would have to be done normally regardless of how many particles you wanted forces on. In principle bonded interactions could benefit, but realizing that benefit might be challenging. You would either have to do extra checks for each interaction, which would probably take as much time as just computing it, or else you would have to build a filtered list of interactions every time the selection changed.

The linked manuscript and discussion focuses on pairwise potentials describing interactions between the selected region (e.g. the region around a ligand) and the whole system.
